### DISCUSSION TITLE
Clarify what's the role/power of Microsoft in the Foundation

### DISCUSSION THREAD

**Discussion by Kryptos-FR** [dotnet-foundation/Home#63] [Board of Directors]
Another day, another drama: https://github.com/dotnet/sdk/issues/22247#issuecomment-949420695.

As far as I understand, all projects under the Foundation umbrella are expected to follow basic open-source guidelines, which implies involving the community before making any big decision.

This was not the case for that PR (https://github.com/dotnet/sdk/pull/22217) that removed a feature that people relied upon and also preventing any debate/discussion by having it locked right away.

If Microsoft cannot play fair with the projects they have "donated" to the Foundation, maybe it's time to end the massacre and stop pretending that they are open. Let's end Microsoft involvement in the Foundation before it takes everyone's sanity. That also means moving all those *fake* open-source projects back to Microsoft's organization on GitHub where they can do whatever they want,w without wasting everybody else's time.

edit: if you are wondering why I am asking that: the PR was locked but the https://github.com/dotnet organization user/admin. Which means while it is supposed to be under the control of the foundation, it was taken over by Microsoft which is not normal.

**Comment by rprouse** [comment]
The .NET Foundation does not control the features that go into or don't go into any of our member projects. In this case, the maintainer is the .NET Team who are employed by Microsoft. That team has every right to decide what makes it into a release and what doesn't just like I do as the NUnit maintainer. There are often people who really want or even need features in NUnit that we don't do because of the complexity it will add to the codebase or because it isn't ready yet.

**Reply by Kryptos-FR** [reply]
It is not the same situation as you described. It was already released in the main branch and was later removed without discussion, that's not the same as accepting/declining to merge a new feature as a maintainer. Or are you saying that as a NUnit maintainer you could decide unilaterally and without discussion to remove a released feature? If so, I might see NUnit with a new pair of eyes.

**Reply by Kryptos-FR** [reply]
Also I could be missing something but the https://github.com/dotnet/ shows the twitter handle/contact email of the .NET foundation, not Microsoft's. So to me it means this organization is not Microsoft, they have their own here: https://github.com/Microsoft/

If I am mistaken, and indeed https://github.com/dotnet/ is a single-vendor organization 100% owned by Microsoft, then it shouldn't pretend to be from the .NET Foundation. This comes back to my original question here: is the Foundation owned by Microsoft or is it really independent.

**Reply by rprouse** [reply]
NUnit has removed features from pre-releases that we didn't feel were ready and we have deprecated and eventually removed many features that were causing issues as have most projects. The point however is that the foundation does not tell its projects what goes into or gets removed from their project. That is not part of our mandate. I understand you don't like the decision and I can sympathize with it.

**Reply by wastaz** [reply]
While I personally at least agree that @rprouse should be allowed to pull whatever feature he wants out of NUnit whenever he wants to, I think my main issue with this argument is that @rprouse is a single maintainer of an open source project with alternatives whereas when it comes to dotnet we are talking about the basic building blocks for the entire dotnet world.

If NUnit does some "questionable" decisions in regard to what features it adds/removes we can switch to xunit or whatever. Or maybe even take a branch and maintain it ourselves. It's an investment, but its not completely out there.

Dotnet however.
Should we switch to java?
I mean, the dotnet team plays in a completely different league here and for completely different stakes. AND on top of this they have the huge microsoft behemot backing them that ofc wants to use dotnet for profit. They need to be seen as responsible caretakers of dotnet. 

If they had taken the entire feature out completely saying "its not ready yet, we are pulling it from the release entirely" then sure. It's sad but it's fine. But taking it out _and sticking it in VS_ is a very different thing.

**Reply by Kryptos-FR** [reply]
I'm trying to raise a more general issue of governance, but you are trying to limit the focus on that specific PR. I feel like you are not replying in good faith, so I will let other people take over the conversation, give their own opinion, and come back later, after reflecting a bit more about it.

**Reply by jmiddour** [reply]
[Moved to outer comment here](https://github.com/dotnet-foundation/Home/discussions/63#discussioncomment-1521726)

**Reply by gortok** [reply]
@rprouse At least for me: I'm not asking for the .NET Foundation to say "this can go or this can stay."

I'm asking the .NET Foundation to stand up for its mission and mandate:

> The .NET Foundation is an independent, non-profit organization established to support an innovative, commercially friendly, open-source ecosystem around the .NET platform. 

By the .NET Foundation staying silent when one of its corporate sponsors specifically detracts from the _innovative_, _commercially friendly_, _open source_, _ecosystem_, it lessens the importance of the .NET Foundation.

If you don't speak up when someone is going directly against your mission, why do you exist as a foundation?

**Comment by matkoniecz** [comment]
> Let's end Microsoft involvement in the Foundation before it takes everyone's sanity.

This seems unlikely, .NET foundation is designed to be Microsoft-controlled.

See https://github.com/dotnet-foundation/Home/discussions/60 for a better overview that I did here by far more experienced person.

In summary, main site of .NET foundation lists "Independent." However, this is a lie.

-----------------------------------

See https://dotnetfoundation.org/about/bylaws

> Section 2.4. Founding Member. Microsoft Corporation is the “Founding Member.”

Microsoft has broad veto rights

> In connection with any vote to materially change the Foundation’s Membership Policy, Director Election Policy, Project Governance Policy, or any Intellectual Property-related agreements or policies, (a) a “no” vote by the director appointed by the Founding Member will result in the disapproval of a proposed action, regardless of the number of votes for approval), and (b) such director must be present as part of any quorum (i.e. if that director is not present, the Board will not have a quorum for the matter, regardless of the number of other directors present).

Especially "any Intellectual Property-related agreements or policies" part can be constructed by sufficiently paid lawyers to cover basically all important activity.

> Any amendment of the Articles of Incorporation or these Bylaws must be approved by vote of two-thirds (2/3) of the directors then in office. Any such amendment that materially alters, reduces or eliminates the rights, responsibilities and privileges of the Founding Member must be agreed to in writing by an authorized representative of the Founding Member who is not serving as a director of the Foundation.

Also, controls at least 1/7 of board in addition to veto rights.

> The Board shall consist of one director appointed by the Founding Member and up to six directors elected by Members

---------------------------------

(it is not wrong or surprising that Microsoft wants veto rights here, but describing this foundation as "Independent" is highly misleading)

**Reply by pkese** [reply]
Who cares if .NET foundation is not designed to be Microsoft-controlled  
when Microsoft is pulling features from .NET regardless of foundation...

**Reply by TechnikEmpire** [reply]
As someone who called all this out early on and was ridiculed by the likes of skeet for it, all I can say is, does anyone want some popcorn?

**Reply by GreenReaper** [reply]
Is the .NET foundation even a valid 501(c)(6) at this point? [According to Wikipedia's summary](https://en.wikipedia.org/wiki/501(c)_organization#501(c)(6)) they are "not organized for profit" and "promote the common business interest" rather than "performing a service for its members". This seems dubious at this point. Even if you did restrict the common benefit to board members (I do not know if this applies to this organization), wouldn't such limits on the board, to the obvious benefit of a specific company, make it ineligible as a nonprofit, and [thus ineligible to use the name "Foundation", at least without other words indicating its actual status](https://app.leg.wa.gov/RCW/default.aspx?cite=23.95.305)? (The [pending update](https://legiscan.com/WA/text/SB5034/id/2223097) to that section appears to be to correct a reference.)

The root of this issue is a corporation misusing nonprofit status as a means to appear something other than it is, ultimately to further its own profit. Someone besides Microsoft must have the standing to end this charade.

**Comment by jmiddour** [comment]
EDIT:  moved to top level comment from nested reply 

I think this might clarify your governance issue @Kryptos-FR ... from  [https://docs.github.com/en/communities/moderating-comments-and-conversations/locking-conversations](https://docs.github.com/en/communities/moderating-comments-and-conversations/locking-conversations)

>Locking a conversation creates a timeline event that is visible to anyone with read access to the repository. However, the username of the person who locked the conversation is only visible to people with write access to the repository. For anyone without write access, the timeline event is anonymized.

![](https://docs.github.com/assets/images/help/issues/anonymized-timeline-entry-for-locked-conversation.png)

>While a conversation is locked, only people with write access and repository owners and collaborators can add, hide, and delete comments.

So, if you _don't_ have write access to the repo, the locking of the pr shows as having come from the owner of the org, even if that is not the account that actually did it.

ETA:  This means that its not only most likely the dotnet account didn't lock the PR, folks who do have write access most likely can't even see the issue you are trying to raise because they see the actual user.

**Reply by filipnavara** [reply]
The PR was locked by the author in case you are interested.

**Reply by jmiddour** [reply]
Thanks for the info.  I guessed it was the author just based on the timeline, but didn't want to make that claim without evidence.

**Reply by Kryptos-FR** [reply]
That's terrible UX/UI. GitHub needs to change that.

**Reply by jmiddour** [reply]
I agree that UX improvements can be made here.  However, I would not call it terrible nor would I say that github needs to make it a priority change.

Here's why:  github has finite resources.  They cannot do everything for everyone.  And with a system as flexible as it is, it is virtually impossible to have a great UX for absolutely every possible scenario.  So they prioritize for the UX for scenarios github is designed to be used for.

based on the design of github, the .Net foundation does not use github the way github expects github to be used.  github is designed for individuals and single organizations to have a place to do source control of assets (either public or private).  github is _not_ designed for a repository fully maintained by a commercial organization (Microsoft) to be "owned" by a nominally independent organization (.Net Foundation).  -- actual independence of the foundation is a separate issue for a different thread.

this UX works absolutely stellar in the scenarios github is designed to be used for.  Changing it to better suit an "off label" use (even one as big as this) runs the risk of degrading the UX in the intended scenarios.

I would put it on the backlog, but I honestly wouldn't touch it until I was already making feature changes to the timeline and was doing a UX eval anyway.  Even then, I'd probably at most link the "locked and limited" text to the "Locking conversation" section of the github docs (which is the first search result for github lock pull request and is how I found the info... didn't know it before hand) similar to how the timeline backlinks mentions.

**Reply by matkoniecz** [reply]
Or maybe modify it to "Account from XYZ".

**Comment by timheuer** [comment]
Hi folks, regarding the note from @Kryptos-FR 
> PR was locked but[sic] the @dotnet organization user/admin. 

Just confirming what @jmiddour notes here...the `dotnet` admin account was not used here, it's a result of how GitHub works apparently.  Here's what non-writers see on the repo:

![image](https://user-images.githubusercontent.com/4821/138507695-1c43e79b-2072-4da6-8e10-69030a3d650c.png)

and here's what write-access sees:
![image](https://user-images.githubusercontent.com/4821/138507723-c38a3590-14ae-4c87-9341-6df5c87497df.png)


I have no idea why that is the case for GitHub features, but just confirming that the `dotnet` account wasn't the one, just that it is the org owner, and that becomes the default in how GitHub works.

**Reply by terrajobst** [reply]
Also "dotnet" is the GitHub organization. In general, a GitHub orgs and user names are in the same namespace, so if there is an org with a name X then there can't be a user with the name X.

**Reply by TheLastCoder** [reply]
welcome @terrajobst . Can you say something about this drama? Or are you not allowed to speak?
(for more details: https://www.theverge.com/2021/10/22/22740701/microsoft-dotnet-hot-reload-removal-decision-open-source )

But congratulations to Microsoft. As much as we want to believe in microsoft, it proves more than once that any product made by microsoft should not be trusted.

**Reply by Kryptos-FR** [reply]
That's terrible UI/UX which create the confusion I was in. I see no reasons for that, GitHub needs to change that.

**Reply by filipnavara** [reply]
@Kryptos-FR The main use case for locking issues and pull requests is to handle abuse or to prevent people from bringing "dead" issues from the graveyard with unrelated comments. In case of the abuse it is totally warranted that the name of the moderator is hidden to prevent further abuse.

In this particular case the issue was locked pre-emptively which is NOT the intended primary use of the tool. I don't think that's GitHub's fault; it's a governance issue.

**Reply by Kryptos-FR** [reply]
I can see the name of the author and all the people that contributed comments to the PR. How exactly is it preventing abuse?

**Reply by filipnavara** [reply]
The design of the locking feature assumes that the abuser is not the same person who has the write access or who authored the PR. (Not judging the particular person here or the events around it, I don't think they are the ones to deserve the criticism.)

**Reply by Kryptos-FR** [reply]
Understood. But it would be better to show a different message than pretend it is the admin account that did it.
Or add a *why am I seeing this (?)* with an underlying link that points to that policy.

My take on this would have been a bit different, had I known that UX behavior.

**Comment by gortok** [comment]
@rprouse It's been almost 6 months and it's been radio silence from the foundation on the topics brought up last October. When can we expect you all to communicate with us?

**Reply by rprouse** [reply]
We are communicating regularly with project maintainers in the Maintainers committee and individually with projects that want to leave GHE or want to leave the Foundation. Those issues are our current priorities.

**Reply by gortok** [reply]
@rprouse Thanks, should I set up a different discussion for the following issues then?

1. Producing governance documents behind the .NET Foundation
2. Claire Novotny is listed as the ED on the .NET Website, and not all the officers in the documents we can find are listed on the site.
3. Executive Director search -- how's that going? We've heard nothing on it.  How can the foundation survive without an ED?

**Reply by rprouse** [reply]
I'll talk to the board about all three. We've been planning on doing 1, we just need to get them uploaded from Sharepoint. 2 is a PR, we can do that, we just keep forgetting. 3, I'll get back to you shortly.

**Reply by matkoniecz** [reply]
@rprouse For start, removing "Independent" from https://dotnetfoundation.org/ or changing bylaws that give Microsoft [broad veto rights](https://github.com/dotnet-foundation/Home/discussions/63#discussioncomment-1520898).

![screen08](https://user-images.githubusercontent.com/899988/156455323-1c5c975b-c959-40e1-9f9c-f7e1400d5698.png)

(it is not wrong or surprising that Microsoft wants veto right here, but describing this foundation as "Independent" is highly misleading)

EDIT: I opened https://github.com/dotnet-foundation/website/pull/1098 implementing removal of this false statement
